pkg/agent/monitoraddnodes.go

```diff
@@ -0,0 +1,272 @@
+const (
+	InitialState addNodeState = iota
+	AssistedServiceIsUp
+	KubeletIsRunningOnNode
+	FirstCSRPendingApproval
+	SecondCSRPendingApproval
+	NodeJoinedClusterStatusNotReady
+	Finish // terminal state
+)
+
+var stateNameMap = map[int]string{
+	int(InitialState):                    "Initial State",
+	int(AssistedServiceIsUp):             "Assisted Service API is available",
+	int(KubeletIsRunningOnNode):          "Kubelet is running",
+	int(FirstCSRPendingApproval):         "First CSR Pending approval",
+	int(SecondCSRPendingApproval):        "Second CSR Pending approval",
+	int(NodeJoinedClusterStatusNotReady): "Node joined cluster and status is NotReady",
+	int(Finish):                          "Node is Ready",
+}
+
+type checkFunction func(nodeMonitor *addNodeMonitor)
+
+type checkFunctions map[int]checkFunction
+
+var defaultChecks = checkFunctions{
+	int(AssistedServiceIsUp): func(mon *addNodeMonitor) {
+		if mon.in(InitialState) &&
+			mon.cluster.API.Rest.IsRestAPILive() {
+			mon.setState(AssistedServiceIsUp)
+		}
+	},
+	int(KubeletIsRunningOnNode): func(mon *addNodeMonitor) {
+		if mon.onOrBefore(KubeletIsRunningOnNode) &&
+			mon.isKubeletRunningOnNode() {
+			mon.setState(KubeletIsRunningOnNode)
+		}
+	},
+	int(FirstCSRPendingApproval): func(mon *addNodeMonitor) {
+		if mon.onOrAfter(KubeletIsRunningOnNode) &&
+			mon.clusterHasFirstCSRPending() {
+			mon.setState(FirstCSRPendingApproval)
+		}
+	},
+	int(SecondCSRPendingApproval): func(mon *addNodeMonitor) {
+		if mon.onOrAfter(KubeletIsRunningOnNode) &&
+			mon.clusterHasSecondCSRPending() {
+			mon.setState(SecondCSRPendingApproval)
+		}
+	},
+	int(NodeJoinedClusterStatusNotReady): func(mon *addNodeMonitor) {
+		hasJoined, isReady, err := mon.nodeHasJoinedClusterAndIsReady()
+		if err != nil {
+			logrus.Debugf("HasNodeJoinedClusterAndIsReady returned err: %v", err)
+		}
+		if hasJoined && !isReady {
+			mon.setState(NodeJoinedClusterStatusNotReady)
+		}
+	},
+	int(Finish): func(mon *addNodeMonitor) {
+		hasJoined, isReady, err := mon.nodeHasJoinedClusterAndIsReady()
+		if err != nil {
+			logrus.Debugf("HasNodeJoinedClusterAndIsReady returned err: %v", err)
+		}
+		if hasJoined && isReady {
+			mon.setState(Finish)
+		}
+	},
+}
+
+// MonitorAddNodes waits for the a node to be added to the cluster
+// and reports its status until it becomes Ready.
+func MonitorAddNodes(cluster *Cluster, nodeIPAddress string) error {
+	parsedIPAddress := net.ParseIP(nodeIPAddress)
+	if parsedIPAddress == nil {
+		return fmt.Errorf("%s is not valid IP Address", nodeIPAddress)
+	}
+
+	timeout := 90 * time.Minute
+	waitContext, cancel := context.WithTimeout(cluster.Ctx, timeout)
+	defer cancel()
+
+	mon := newAddNodeMonitor(nodeIPAddress, cluster)
+
+	wait.Until(func() {
+		if mon.cluster.API.Rest.IsRestAPILive() {
+			_, err := cluster.MonitorStatusFromAssistedService()
+			if err != nil {
+				logrus.Warnf("Node %s: %s", nodeIPAddress, err)
+			}
+		}
+
+		for _, checkFunc := range defaultChecks {
```

Comment:
I reviewed/debugged this part several times, and I think it could be simplified further. The main issue I found is the fact that iterating on a map does not guarantee the order, so the evaluation of the various checks is not deterministic, and for this reason each check function must perform an additional check to evaluate in which point the state is currently is - making the code more difficult to read and understand.
So, given that the number of states is very limited, and moreover a definitely linear graph, I would find a lot easier a simpler code where the different conditions are evaluated one after the other, at each loop (and not all togheter), ie:

```
// Assisted service was found, so we are still before the first reboot
if mon.if mon.cluster.API.Rest.IsRestAPILive() {
  ...
 return false
}

// Assisted was not found, so either we are still looking at an installation in progress or maybe not (the node already joined)
if mon.nodeHasJoinedClusterAndIsReady() {
  ...
  return true // done
}

if mon.
if mon.clusterHasSecondCSRPending {
  return false
}

if mon.clusterHasFirstCSRPending {
  return false
}

//No csrs found, check if node it's really installing
if mon.isKubeletRunningOnNode() {
  return false
}

//At this point the node could be either between the first reboot and kubelet startup, or something went horribly wrong. Just wait
return false
```

This approach (similar to the one already used in the other wait-for command) would not require a sm at all, and could be easier to read / maintain

Reply:
I've updated it to not use the map and just iterate over the different checks. Hopefully it is easier to read now.